### PR TITLE
Always copy function args into allocas to allow better optimizations

### PR DESCRIPTION
**Pull request by dotdash**
Not copying function arguments into allocas prevents certain
optimizations from happening, for example SROA splitting aggregates into
their scalar components.

For example this function:

```
struct S { x: i64, y: i64, z: Box<i64> }

fn test(s: S) -> S {
  let s = s;
  let s = s;
  let s = s;
  s
}
```

Currently isn't optimized down to a single memcpy, but it is optimized
to that with this change.

To get the full effect, we also need to stop marking the argument as
dead (lifetime.end intrinsic) after copying it into the alloca, because
that also stops certain optimizations from happening. Dropping that
intrinsic comes without negative side effects, as non-allocas don't
participate in stack coloring anyway.

### CODE REVIEW

File: src/librustc_trans/trans/_match.rs [hunk 1]
```diff
@@ -1624,21 +1624,9 @@ pub fn store_arg<'blk, 'tcx>(mut bcx: Block<'blk, 'tcx>,
         Some(ident) => {
             // Generate nicer LLVM for the common case of fn a pattern
             // like `x: T`
```
Comment (alexcrichton) [inline src/librustc_trans/trans/_match.rs]: Is this comment still applicable?

**Comment by alexcrichton** [comment]
I do find it fascinating that doing the "not super smart" thing is causing LLVM to optimize *better* in this case! I feel like this is a prime example where we really need some form of codegen test to make sure this doesn't regress, but alas!

Nice wins!

File: src/librustc_trans/trans/_match.rs (hunk 1 above)
Comment (dotdash) [inline src/librustc_trans/trans/_match.rs]: I understood this to be about the Some/None arms in the surrouding match, simple pattern vs. generic pattern.

**Comment by dotdash** [comment]
Yeah, I only discovered this by chance actually, but thinking about it, it makes sense that optimizations are easier to apply if there's an alloca in between. We'll later also want to do this for return slots, that allows e.g. derived clone implementations to be far better optimized, but currently, the post-drop zeroing interferes so we get worse code in too many places to make it worth it right now.

File: src/librustc_trans/trans/_match.rs (hunk 1 above)
Comment (alexcrichton) [inline src/librustc_trans/trans/_match.rs]: Ah ok (just making sure)

File: src/librustc_trans/trans/_match.rs [hunk 2]
```diff
@@ -1624,21 +1624,9 @@ pub fn store_arg<'blk, 'tcx>(mut bcx: Block<'blk, 'tcx>,
...
-            let arg_ty = node_id_type(bcx, pat.id);
-            if type_of::arg_is_indirect(bcx.ccx(), arg_ty)
-                && bcx.sess().opts.debuginfo != FullDebugInfo {
-                // Don't copy an indirect argument to an alloca, the caller
-                // already put it in a temporary alloca and gave it up, unless
-                // we emit extra-debug-info, which requires local allocas :(.
```
Comment (eddyb) [inline src/librustc_trans/trans/_match.rs]: So is this PR a trade-off, or did LLVM get better than it was back when this comment was written?
Also, if we always copy, that means we don't need to create new copies in the callers - which should always be an improvement if the callee copies aren't mutated.

IMO, we should never copy a value that moves, even if there are mutations, and I'm not sure we can teach LLVM about what is effectively `&move T`.

**Comment by dotdash** [comment]
@bors r-

Trying to experiment with not copying incoming args in the caller, I noticed some problems originating from this change. I should have a deeper look at this before we merge this.

**Comment by dotdash** [comment]
This causes trouble when the address of the argument (or any of its fields in case of an aggregate) is taken.

**Pull request closed without merging**